File: server/src/service/message.service.ts
import { Provide, Inject } from "@midwayjs/core";
import { PrismaClient, MessageType } from "@prisma/client";
import { PAGINATION } from "../common/constants";

@Provide()
export class MessageService {
  @Inject("prisma")
  prisma!: PrismaClient;

  /** 创建消息 */
  async create(
    userId: number,
    type: MessageType,
    title: string,
    content: string,
  ) {
    return this.prisma.message.create({
      data: {
        userId,
        type,
        title,
        content,
      },
    });
  }

  /** 消息列表 */
  async findByUser(
    userId: number,
    type?: MessageType,
    isRead?: boolean,
    page = PAGINATION.DEFAULT_PAGE,
    pageSize = PAGINATION.DEFAULT_PAGE_SIZE,
  ) {
    const where = {
      userId,
      ...(type && { type }),
      ...(isRead !== undefined && { isRead }),
    };

    const [list, total] = await Promise.all([
      this.prisma.message.findMany({
        where,
        orderBy: { createdAt: "desc" },
        skip: (page - 1) * pageSize,
        take: pageSize,
      }),
      this.prisma.message.count({ where }),
    ]);

    return {
      list,
      total,
      page,
      pageSize,
      totalPages: Math.ceil(total / pageSize),
    };
  }

  /** 标记已读 */
  async markAsRead(messageId: number, userId: number) {
    const message = await this.prisma.message.findUnique({
      where: { id: messageId },
    });

    if (!message) {
      throw new Error("消息不存在");
    }
    if (message.userId !== userId) {
      throw new Error("无权操作此消息");
    }

    return this.prisma.message.update({
      where: { id: messageId },
      data: { isRead: true },
    });
  }

  /** 全部标记已读 */
  async markAllAsRead(userId: number) {
    return this.prisma.message.updateMany({
      where: {
        userId,
        isRead: false,
      },
      data: { isRead: true },
    });
  }

  /** 未读消息数 */
  async getUnreadCount(userId: number): Promise<number> {
    return this.prisma.message.count({
      where: {
        userId,
        isRead: false,
      },
    });
  }
}

File: server-java/src/main/java/com/wudong/service/MessageService.java
package com.wudong.service;

import com.wudong.common.enums.MessageType;
import com.wudong.entity.Message;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * 消息服务 - 等价于 Node.js 版本的 MessageService
 */
public interface MessageService {

    /**
     * 创建消息
     */
    Message createMessage(Long userId, MessageType type, String title, String content);

    /**
     * 获取消息列表（支持组合过滤）
     */
    Page<Message> getMessages(Long userId, MessageType type, Boolean isRead, Pageable pageable);

    /**
     * 标记单条消息已读
     */
    void markAsRead(Long messageId, Long userId);

    /**
     * 标记所有消息已读
     */
    void markAllAsRead(Long userId);

    /**
     * 获取未读消息数
     */
    long getUnreadCount(Long userId);
}
